packages/server/lib/util/agent.ts

```diff
@@ -0,0 +1,259 @@
+import * as http from 'http'
+import * as https from 'https'
+import * as net from 'net'
+import * as tls from 'tls'
+import * as url from 'url'
+import * as debugModule from 'debug'
+import { getProxyForUrl } from 'proxy-from-env'
+import * as Promise from 'bluebird'
+import * as connect from './connect'
+
+const debug = debugModule('cypress:server:agent')
+
+function createProxySock (proxy: url.Url) {
+  if (proxy.protocol === 'http:') {
+    return net.connect(Number(proxy.port || 80), proxy.hostname)
+  }
+
+  if (proxy.protocol === 'https:') {
+    // if the upstream is https, we need to wrap the socket with tls
+    return tls.connect(Number(proxy.port || 443), proxy.hostname)
+  }
+
+  // socksv5, etc...
+  throw new Error(`Unsupported proxy protocol: ${proxy.protocol}`)
+}
+
+function regenerateRequestHead(req) {
+  req._header = null;
+  req._implicitHeader();
+  if (req.output && req.output.length > 0) {
+    // the _header has already been queued to be written to the socket
+    var first = req.output[0];
+    var endOfHeaders = first.indexOf('\r\n\r\n') + 4;
+    req.output[0] = req._header + first.substring(endOfHeaders);
+  }
+}
+
+export class CombinedAgent {
+  httpAgent: HttpAgent
+  httpsAgent: HttpsAgent
+  familyCache = {}
+
+  constructor(httpOpts: http.AgentOptions = {}, httpsOpts: https.AgentOptions = {}) {
+    this.httpAgent = new HttpAgent(httpOpts)
+    this.httpsAgent = new HttpsAgent(httpsOpts)
+  }
+
+  // called by Node.js whenever a new request is made internally
+  addRequest(req: http.ClientRequest, options: any) {
+    // WSS connections will not have an href, but you can tell protocol from the defaultAgent
+    const isHttps = (options._defaultAgent && options._defaultAgent.protocol === 'https:')
```

Comment:
let's extract this into a function with a name that helps convey what this is doing